### ISSUE TITLE
BUG: CategoricalIndex.get_indexer with 

### ISSUE THREAD

**Issue by jbrockmendel** [pandas-dev/pandas#45361]
```python
ci = pd.CategoricalIndex([1, 2, np.nan, 3])
other = pd.Index([2, 3, 4])

res = ci.get_indexer(other)

>>> res
array([1, 3, 2])
```

The 4 in `other` is getting mapped to the `nan` in `ci`.  Best guess is that this is passing `other` to the Categorical constructor which will return `Categorical([2, np.nan, 3], dtype=ci.dtype)`.  If correct, this would be avoided by #40996.

### Expected Behavior

```
>>> res
array([1, -1, 2])
```

**Comment by Shashank-Shet** [comment]
@jbrockmendel   Shouldn't expected output be `array([1, 2, -1])`

**Comment by Shashank-Shet** [comment]
I think I understand what is going on here. Internally, `[2, 3, 4]` is treated as `[2, 3, nan]` since 4 is not a category in `ci`.
So instead of 4 being mapped to -1, the nan in its place is mapped to 2.

**Comment by jbrockmendel** [comment]
PR would be welcome!

**Comment by Shashank-Shet** [comment]
I have just submitted a PR.

**Comment by Shashank-Shet** [comment]
I just needed a clarification on this piece of code:
```python
ci = pd.CategoricalIndex([1,2,3,np.nan])
ci.get_indexer([1,2,4,np.nan])
```
Should the output be `[0,1,-1,-1]`? Or `[0,1,-1,3]`?

**Comment by jbrockmendel** [comment]
closed by #45373
